alf/optimizers/utils.py

```diff
@@ -116,11 +121,10 @@ def forward(self, loss: torch.Tensor, tensors: alf.nest.NestedTensor):
                 represents an individual loss on a single training sample. Ideally,
                 the samples used for computing these losses should be sampled *with*
                 replacement independently. The loss can have a shape of either
-                ``[T,B]`` or ``[B]``. In the first case, we will only take
-                ``[0,B]`` for computing the gns due to sample correlation along
-                the temporal axis. On-policy algorithms might have noisy estimates
-                if the ``num_parallel_environments`` is small. Similarly, this
-                happens for a small batch size for off-policy algorithms.
+                ``[T,B]`` or ``[B]``. In the first case, we will randomly select
+                a time step ``t``. The reason is that if the ``T`` steps' gradients
+                are consider together, it will violate the i.i.d. assumption of
```

Comment:
Treating each sequence as a sample, the samples of a batch are still IID.
And using a single t does not work for RNN.

Reply:
> Treating each sequence as a sample, the samples of a batch are still IID. And using a single t does not work for RNN.

I think as long as the subsequence is randomly sampled with a random start position, using a single t still works. But so far it turns out neither is very stable.

Reply:
> Treating each sequence as a sample, the samples of a batch are still IID. And using a single t does not work for RNN.

In fact, there are multiple losses computed along a sequence, each loss for a time step. Even though sequences are iid with each other, the losses are not from iid samples?